source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst

```diff
@@ -355,29 +397,85 @@ Before a connection can be returned outside of the pool, it must be “set up”
 Closing a Connection (Internal Implementation)
 ----------------------------------------------
 
-When a Connection is closed, it MUST first be marked as closed, removing it from being counted as available, in use, or setup. Once the connection is marked as closed, the Connection can perform whatever teardown is necessary. The Driver SHOULD perform the teardown in a non-blocking manner.
+When a `Connection <#connection>`_ is closed, it MUST first be marked as "closed",
+removing it from being counted as "available" or "in use". One that is
+complete, the `Connection <#connection>`_ can perform whatever teardown is
+necessary to close its underlying socket. The Driver MUST perform this
+teardown in a non-blocking manner, such as via the use of a background
+thread or async I/O.
 
 .. code::
 
-    mark connection as closed
+    decrement total connection count
+    if connection state is "available":
+      decrement available connection count
+    set connection state to "closed"
     emit ConnectionClosedEvent
 
-    # The following can happen at a later time
-    connection.close()
+    # The following can happen at a later time (i.e. in background
+    # thread) or via non-blocking I/O.
+    connection.socket.close()
+
+Marking a Connection as Available (Internal Implementation)
+-----------------------------------------------------------
+
+A `Connection <#connection>`_ is "available" if it is able to be checked out. A
+`Connection <#connection>`_ MUST NOT be marked as "available" until it has been
+established. The pool MUST keep track of the number of currently
+available `Connections <#connection>`_.
+
+.. code::
+
+   increment available connection count
+   set connection state to "available"
+   add connection to availableConnections
+
+
+Populating the Pool with a Connection (Internal Implementation)
+---------------------------------------------------------------
+
+If minPoolSize is set, the pool MUST ensure that it always has at
+least minPoolSize total `Connections <#connection>`_. In order to achieve this, it
```

Comment:
The wording earlier in the spec ("the pool MUST immediately begin populating") was used to reflect the fact that connections are not created instantly. A pool that provides a minimum number of *usable* connections cannot *always* have that many connections.

A pool that provides any number of dummy connection objects can make them as soon as the pool is constructed. This is how the Ruby driver worked prior to 2.11. We do not consider this useful behavior because connection objects are lightweight - if the driver does not understand min pool size to mean that many *usable* connections, it's pointless to have a min pool size at all.

For this hunk, I suggest not requiring the min pool size with a MUST. Either omitting this sentence entirely or repeating "the pool MUST immediately begin populating" version could be viable alternatives.